dd-sdk-android-ndk/src/main/cpp/datadog-native-lib.cpp

```diff
@@ -0,0 +1,112 @@
+static const char *ERROR_KIND = "Native";
+static const char *LOGGER_NAME = "crash";
+static const char *LOG_TAG = "NdkCrashReporter";
+static const char *SERVICE_NAME = "";
+static const char *STORAGE_DIR = nullptr;
+
+std::string get_serialized_log(int signal,
+                               const char *signal_name,
+                               const char *error_message,
+                               const char *date,
+                               const char *backtrace) {
+    std::string serializedLog;
+    serializedLog.append("{ ");
+    serializedLog.append(R"("message": ")").append(GENERIC_MESSAGE).append("\",");
+    serializedLog.append(R"("service": ")").append(SERVICE_NAME).append("\",");
+    serializedLog.append(R"("logger.name": ")").append(LOGGER_NAME).append("\",");
+    serializedLog.append(R"("env": ")").append(ENVIRONMENT).append("\",");
+    serializedLog.append(R"("status": ")").append(EMERGENCY_STATUS).append("\",");
+    serializedLog.append(R"("date": ")").append(date).append("\",");
+    serializedLog.append(R"("error.stack": ")").append(backtrace).append("\",");
+    serializedLog.append(R"("error.message": ")").append(error_message).append("\",");
+    char formattedSignalMessage[30];
+    sprintf(formattedSignalMessage, "%s: %d", signal_name, signal);
+    serializedLog.append(R"("error.signal": ")").append(formattedSignalMessage).append("\",");
+    serializedLog.append(R"("error.kind": ")").append(ERROR_KIND).append("\"");
+    serializedLog.append(" }");
+    return serializedLog;
+}
+
+void crash_signal_intercepted(int signal, const char *signal_name, const char *error_message) {
+    // sync everything
+    static pthread_mutex_t handlerMutex = PTHREAD_MUTEX_INITIALIZER;
+    pthread_mutex_lock(&handlerMutex);
+
+    if (STORAGE_DIR == nullptr) {
+        __android_log_write(ANDROID_LOG_ERROR, LOG_TAG,
+                            "The crash reports storage directory file path was null");
+        pthread_mutex_unlock(&handlerMutex);
+        return;
+
+    }
+
+    // create crash reporting directory if it does not exist
+    if (!fileutils::createDirIfNotExists(STORAGE_DIR)) {
+        __android_log_print(ANDROID_LOG_ERROR, LOG_TAG,
+                            "Was unable to create the NDK reports storage directory: %s",
+                            STORAGE_DIR);
+        pthread_mutex_unlock(&handlerMutex);
+        return;
+    }
+
+    // format the current GMT time
+    char date[80];
+    const char *format = "%Y-%m-%d'T'%H:%M:%S.000Z";
+    date_with_format(date, sizeof(date), format);
+
+    // extract the generate_backtrace
+    const char *backtrace = trace::generate_backtrace();
+
+    // serialize the log
+    std::string serialized_log = get_serialized_log(signal, signal_name, error_message, date,
+                                                    backtrace);
+
+    // dump the log into a new file
+    char filename[200];
+    sprintf(filename, "%s/%llu", STORAGE_DIR, time_since_epoch());
```

Comment:
OK, I told you something wrong. `sprintf()` is not safe, you should use `snprintf()` and pass the size of your buffer.

Reply:
No worries...good to know ;)

Reply:
It is done already